docs: new tutorial "generate code" (#1025)
asyncapi/website: pages/docs/tutorials/generate-code.md

The generator tool supports a number of templates to generate code for a variety of different languages and protocols as the output. These templates help to specify what exactly must be generated, and in this tutorial, you'll use a [Node.js template](https://github.com/asyncapi/nodejs-template).

## Installation guide
<Remember>

It is important to have [Node.js](https://nodejs.org/en/download/) installed which will enable the installation of the required dependencies using <b>npm</b>.

</Remember>

import CliInstallation from '../../../assets/docs/fragments/cli-installation.md' 

<CliInstallation/>

## Generate code

To generate code from the [AsyncAPI document created in a previous tutorial](https://asyncapi.com/docs/tutorials/create-asyncapi-document), follow the steps listed below:

<Remember>

If you did not follow the previous tutorial and do not have an `asyncapi.yaml` file ready, generate one running `asyncapi new --example=tutorial.yml --no-tty`.

</Remember>

> 1. Trigger generation of the Node.js code:
> <CodeBlock language="bash">
> {`asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto`}
> </CodeBlock>
> 
> Let's break down the previous command:
> - `asyncapi generate fromTemplate` is how you use AsyncAPI Generator via the AsyncAPI CLI. 
> - ` asyncapi.yaml` is how you point to your AsyncAPI document and can be a URL. 
> - `@asyncapi/nodejs-template` is how you specify the Node.js template.
> - `-o` determines where to output the result.
> - `-p` defines additional parameters you want to pass to the template. Here, the `server` parameter specifies the server's name as it is defined in AsyncAPI document.

Reviewer: ```suggestion
1. Trigger generation of the Node.js code:
    <CodeBlock language="bash">
    {`asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto`}
    </CodeBlock>
    Let's break down the previous command:
    - `asyncapi generate fromTemplate` is how you use AsyncAPI Generator via the AsyncAPI CLI. 
    - ` asyncapi.yaml` is how you point to your AsyncAPI document and can be a URL. 
    - `@asyncapi/nodejs-template` is how you specify the Node.js template.
    - `-o` determines where to output the result.
    - `-p` defines additional parameters you want to pass to the template. Here, the `server` parameter specifies the server's name as it is defined in AsyncAPI document.
```

<CodeBlock language="bash">
{`cd output && ls`}
</CodeBlock>

Upon execution of the command above, the following is an example of the expected result:
```
$ ls
Dockerfile
asyncapi.yaml
docs
src
README.md
config
package.json
```

## Start generated application
1. Install dependencies of the newly generated application:
<CodeBlock language="bash">
{`npm install`}
</CodeBlock>

2. Start the application:
<CodeBlock language="bash">
{`npm start`}
</CodeBlock>

## Send message to broker
1. In another terminal install the MQTT.js library:
<CodeBlock language="bash">
{`npm install mqtt -g`}
</CodeBlock>

2. Send a message to the MQTT broker that's connected to your generated application. Run the following MQTT client command:
<CodeBlock language="bash">
{`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
</CodeBlock>

3. Go back to the previous terminal to check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal:
    ```bash
    light/measured was received:
    { id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }

>     ```

Reviewer: ```suggestion
1. In another terminal install the MQTT.js library:
    <CodeBlock language="bash">
{`npm install mqtt -g`}
    </CodeBlock>

2. Send a message to the MQTT broker that's connected to your generated application. Run the following MQTT client command:
    <CodeBlock language="bash">
    {`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
    </CodeBlock>

3. Go back to the previous terminal to check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal:
    ```bash
    light/measured was received:
    { id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }
    ```
```

> 3. Go back to the previous terminal to check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal:
> <CodeBlock language="bash">
> light/measured was received:
> { id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }
> </CodeBlock>

Reviewer: ```suggestion
3. Go back to the previous terminal to check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal:
    ```bash
    light/measured was received:
    { id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }
    ```
```

I think it is some bug with `CodeBlock`. I suggest you log separate issue with failing example. In meantime use "```". Formatting works well if you indent code block under the number list with 4 spaces

---
title: Generate code 
description: In this tutorial, you'll learn how to generate code from your AsyncAPI document.
weight: 100
---

## Introduction

In this tutorial, you'll learn how to generate code from your AsyncAPI document using the AsyncAPI generator tool.

## Background context
The [AsyncAPI Generator](https://github.com/asyncapi/generator) is a tool that you can use to generate whatever you want based on the AsyncAPI document. You can generate docs and code. It can be used as a library in Node.js application or through the [AsyncAPI CLI](https://github.com/asyncapi/cli).

The generator tool supports a number of templates to generate code for a variety of different languages and protocols as the output. These templates help to specify what exactly must be generated, and in this tutorial, you use [Node.js template](https://github.com/asyncapi/nodejs-template).

## Installation guide
<Remember>

> It is important to have <a href='https://nodejs.org/en/download/'>Node.js</a> installed which will enable the installation of the required dependencies using <b>npm</b>.

Reviewer 2: ```suggestion
It is important to have [Node.js](https://nodejs.org/en/download/) installed which will enable the installation of the required dependencies using <b>npm</b>.
```

</Remember>

> - Install the generator tool to use it as a command-line (CLI):
> 
> <CodeBlock language="bash">
> {`npm install -g @asyncapi/cli`}
> </CodeBlock>
> 
> The command above helps you install the generator tool globally.
> 
> - To install a specific version of the generator tool, pass the version during installation:
> 
> <CodeBlock language="bash">
> {`npm install -g @asyncapi/cli@{version}`}
> </CodeBlock>

Reviewer: ```suggestion
```

this is a duplication of what you bring in with the CLI installation fragment document

import CliInstallation from '../../../assets/docs/fragments/cli-installation.md' 

<CliInstallation/>

## Generate code

> To generate code from the AsyncAPI document created from the [previous tutorial](""), follow the steps listed below:

Reviewer 2: Why the link `previous tutorial` has been left blank?

<CodeBlock language="bash">
{`cd output && ls`}
</CodeBlock>

Upon execution of the command above, the following is an example of the expected result;
```
$ ls
Dockerfile
asyncapi.yaml
docs
src
README.md
config
package.json
```

## Start generated application

1. Install dependencies of newly generated application:
<CodeBlock language="bash">
{`npm install`}
</CodeBlock>

2. Start the application:
<CodeBlock language="bash">
{`npm start`}
</CodeBlock>

## Send message to broker

1. In another terminal install the MQTT.js library:
<CodeBlock language="bash">
{`npm install mqtt -g`}
</CodeBlock>

2. Send a message to MQTT broker that the generated application is now connected to. You will use the MQTT client command in your terminal:
<CodeBlock language="bash">
{`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
</CodeBlock>

3. Go back to the previous terminal and check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal:

```
light/measured was received:
{ id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }
```

## Summary

> In this tutorial, you learned how to generate your code from the Streetlights API specification document created in the [previous tutorial]() using the AsyncAPI generator tool. 

Reviewer 2: Here also, the `previous tutorial` link is kept blank.

Additionally, you have learned how to run your code by installing the generated code's dependencies and sending several test messages to the Streelights application using the MQTT client.

## Next steps

> Now that you've completed this tutorial, go ahead to learn how to [validate your document using the AsyncAPI studio]() through the validation techniques supported by AsyncAPI.

Reviewer 2: Kindly add the link to the tutorial mentioned here.

The generator tool supports a number of templates to generate code for a variety of different languages and protocols as the output. These templates help to specify what exactly must be generated, and in this tutorial, you use [Nodejs template](https://github.com/asyncapi/nodejs-template).

## Installation guide
<Remember>

It is important to have <a href='https://nodejs.org/en/download/'>Node.js</a> installed which will enable the installation of the required dependencies using <b>npm</b>.

</Remember>

- Install the generator tool to use it as a command-line (CLI):

<CodeBlock language="bash">
{`npm install -g @asyncapi/cli`}
</CodeBlock>

The command above helps you install the generator tool globally.

- To install a specific version of the generator tool, pass the version during installation:

<CodeBlock language="bash">
{`npm install -g @asyncapi/cli@{version}`}
</CodeBlock>

## Generate code

> To generate code from the AsyncAPI document created from the [previous tutorial](""), follow the steps listed below;

Reviewer: I'm not sure if you followed convo in slack. So basically @alequetzalli goal is to have tutorials standalone, and best if they do not refer to each other this way.

so basically write something like a note where you specify, if you did not follow previous tutorials and do not have `asyncapi.yaml` file ready, then generate one using `asyncapi new --example=tutorial.yml --no-tty` command

- `-p` defines additional parameters you want to pass to the template, in this instance, the `server` parameter which is where you specify the server's name as it is defined in AsyncAPI document.

2. List all files in directory and check that Node.js application is generated:
<CodeBlock language="bash">
{`cd output && ls`}
</CodeBlock>

Upon execution of the command above, the following is an example of the expected result;
```
$ ls
Dockerfile
asyncapi.yaml
docs
src
README.md
config
package.json
```

## Start generated application

1. Install dependencies of newly generated application:
<CodeBlock language="bash">
{`npm install`}
</CodeBlock>

2. Start the application:
<CodeBlock language="bash">
{`npm start`}
</CodeBlock>

## Send message to broker

1. In another terminal install the MQTT.js library:
<CodeBlock language="bash">
{`npm install mqtt -g`}
</CodeBlock>

2. Send a message to MQTT broker that the generated application is now connected to. You will use the MQTT client command in your terminal:
<CodeBlock language="bash">
{`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
</CodeBlock>

> 3. Go back to the previous terminal and check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal;

Reviewer: ```suggestion
3. Go back to the previous terminal and check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal:
```

```
light/measured was received:
{ id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }
```

> # Summary

Reviewer: ```suggestion
## Summary
```

In this tutorial, you learned how to generate your code from the Streetlights API specification document created in the [previous tutorial]() using the AsyncAPI generator tool. 

Additionally, you have learned how to run your code by installing the generated code's dependencies and sending several test messages to the Streelights application using the MQTT client.

> # Next steps

Reviewer: ```suggestion
## Next steps
```

Now that you've completed this tutorial, go ahead to learn how to [validate your document using the AsyncAPI studio]() through the validation techniques supported by AsyncAPI.

---

<DocsButton
  suggestions={[
    {

>       href: '/docs/tutorials/validate-documents.md',

Reviewer: ```suggestion
      href: '/docs/tutorials/validate-documents',
```

      title: 'Validate AsyncAPI document',
      type:'back',
    },
    {

>       href: '/docs/tutorials/message-validation.md',

Reviewer: ```suggestion
      href: '/docs/tutorials/message-validation',
```

---
title: Generate code 
description: In this tutorial, you'll learn how to generate code from your AsyncAPI document.
weight: 130
---

## Introduction

In this tutorial, you'll learn how to generate code from your AsyncAPI document using the AsyncAPI generator tool.

## Background context
The [AsyncAPI Generator](https://github.com/asyncapi/generator) is a tool that you can use to generate whatever you want based on the AsyncAPI document. You can generate docs and code. It can be used as a library in Nodejs application or through the [AsyncAPI CLI](https://github.com/asyncapi/cli).

The generator tool supports a number of templates to generate code for a variety of different languages and protocols as the output. These templates help to specify what exactly must be generated, and in this tutorial, you use [Nodejs template](https://github.com/asyncapi/nodejs-template).

> ## Installation guide

Reviewer: we need more than just node/npm instructions, I think I already pointed you to https://github.com/asyncapi/generator/pull/818/files#diff-6c85f832c3eae8efc6830b5e9e56020250a891fa7fd58498a8506c4600ab3c34R32-R53 but I'm might mix PRs 😄 

bigger question related to it and convo from https://asyncapi.slack.com/archives/C02QY9FMM18/p1668535611547909

cc @alequetzalli 

since tutorials should also be used as standalone, then it means `CLI installation` instructions will show up in more than one tutorial.

I strongly suggest we start reusing content. I mean having one file where this CLI installation is described, and then we include it in as many other documents as needed.

It is technically possible as we already had a convo about it with @magicmatatjahu in this PR https://github.com/asyncapi/website/pull/886#issuecomment-1222760522

I also tested it now quickly on local.
tl;dr it is possible out of the box

VSCode 👇🏼 

![Screenshot 2022-11-16 at 13 28 35](https://user-images.githubusercontent.com/6995927/202180896-5e25a2e0-b1e5-4740-8421-a50ca9109f3f.png)

Result in UI 👇🏼 

![Screenshot 2022-11-16 at 13 29 22](https://user-images.githubusercontent.com/6995927/202181044-7aa8ee3e-9087-4db1-9774-42be76a05d1d.png)

Reviewer 3: Remember that we shouldn't write fragments/reusable content in the `pages/**` folder, because then NextJS will treat (as in your example) the `pages/docs/fragments/**` as separate pages in website. We should move them in the `{root}/docs` folder or something like.

cc @derberg @alequetzalli

Reviewer 4: **Cool! I love the idea of having these reusable components!**

1) I don't think we need to mention the Generator Tool tho? That seemed like an example only relevant to that one tutorial...
2) Are you volunteering to create this? 😂 I am asking just so I follow as to when you hope to start implementing this. Is this something that can be easily/quickly done before this month ends with GSoD program? 🤔

Reviewer 4: nvm... resolved via https://github.com/asyncapi/website/pull/1113

Author: > we need more than just node/npm instructions, I think I already pointed you to https://github.com/asyncapi/generator/pull/818/files#diff-6c85f832c3eae8efc6830b5e9e56020250a891fa7fd58498a8506c4600ab3c34R32-R53 but I'm might mix PRs 😄

Hi @derberg, 
In having more than the node/npm instructions, do you mean adding installation instructions for other operating systems as shown in the link you pointed above.

I remember you once pointed sticking to one OS here 
<img width="558" alt="convo" src="https://user-images.githubusercontent.com/52763841/202599055-552efe62-beb6-4c08-85b7-9a39475a38ef.PNG">

Reviewer: @Annysah we basically need to wait for PR from @alequetzalli to get merged so you can reuse her CLI installation instructions here

Author: Got it!

Reviewer 4: "To generate code, you need to install the AsyncAPI CLI that allows you to generate code..."

> Before proceeding to the next step, 

Reviewer 4: ```suggestion
```

This seems unnecessary

<Remember>

> It is important to have [Node.js](https://nodejs.org/en/download/) installed which will enable the installation of the required dependencies using <b>npm</b>.

Reviewer 4: Heyooo @Annysah...😄✌🏽  So you can't use markup syntax inside a **Remember** `react component`. 

You need to update the link markup to have approved HTML `<a>` tags.

</Remember>

> - Install the generator tool to use it as a command-line(CLI);

Reviewer 4: ```suggestion
- Install the generator tool to use it as a command-line (CLI):
```

{`asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto`}
</CodeBlock>

From the command above;

- `asyncapi generate fromTemplate` is how you use AsyncAPI Generator through the AsyncAPI CLI. 

- ` asyncapi.yaml` is how you point to your AsyncAPI document. It can also be a URL. 

- `@asyncapi/nodejs-template` is how you specify the Node.js template.
 
- `-o` determines where to output the result.

- `-p` defines additional parameters you want to pass to the template, in this instance, the `server` parameter which is where you specify the server's name as it is defined in AsyncAPI document.

2. List all files in directory and check that Node.js application is generated:
<CodeBlock language="bash">
{`cd output && ls`}
</CodeBlock>

Upon execution of the command above, the following is an example of the expected result;
```
$ ls
Dockerfile
asyncapi.yaml
docs
src
README.md
config
package.json
```

## Start generated application

1. Install dependencies of newly generated application:
<CodeBlock language="bash">
{`npm install`}
</CodeBlock>

2. Start the application:
<CodeBlock language="bash">
{`npm start`}
</CodeBlock>

> ## Send a message to the broker

Reviewer 4: ```suggestion
## Send message to broker
```

we can def shorten this to make the TOC more readable 😄

</CodeBlock>

3. Go back to the previous terminal and check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal;

```
light/measured was received:
{ id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }
```

# Summary

In this tutorial, you learned how to generate your code from the Streetlights API specification document created in the [previous tutorial]() using the AsyncAPI generator tool. 

Additionally, you have learned how to run your code by installing the generated code's dependencies and sending several test messages to the Streelights application using the MQTT client.

# Next steps
Now that you've completed this tutorial, go ahead to learn how to [validate your document using the AsyncAPI studio]() through the validation techniques supported by AsyncAPI.

---

<DocsButton
  suggestions={[
    {
      href: '/docs/tutorials',
      title: 'Tutorials - Overview',
      type:'back',
    },
    {
      href: '/docs/tutorials/streetlights-interactive',
      title: 'Streetlights - Interactive (Alpha)',
      type:'next',
    }
  ]}

> />

Reviewer 4: Hey @Annysah, this is wrong. 😄✌🏽 Why do you have prev/next buttons that link to a streetlights tutorial that we are removing? Remember how this PR replaces the streetlights tutorial? (including the interactive one) Second, this tutorial will **not** go after the `Tutorials Overview` page, this tutorial is the 3rd item in line. 

Remember, the 4 new tutorials to replace Streetlights tutorial will go in this order:
1. create AsyncAPI document
2. validate AsyncAPI document
3. generate code
4. validate messages/events 

**Please update this PR and fix the prev/next buttons to link to the correct order of tutorials.**

Author: oh yes, I left it because I was uncertain about how to link it correctly.
will this be a right link for the prev button '/docs/tutorials/validate-documents.md' and '/docs/tutorials/message-validation.md' for the next button?

Reviewer 4: yes, exactly right @Annysah 😄✌🏽 

go forth and update the buttons 🙌🏽🙌🏽🙌🏽

Author: Thank you!...updating right away

---
title: Generate Code 

> description: In this tutorial, you'll learn how to generate your code from an AsyncAPI document.

Reviewer: ```suggestion
description: In this tutorial, you'll learn how to generate code from your AsyncAPI document.
```

weight: 130
---

# Introduction

In this tutorial, you'll learn how to generate your code from an AsyncAPI document using the AsyncAPI generator tool.

# Background Context
The [AsyncAPI Generator](https://github.com/asyncapi/generator) is a tool that you can use to generate whatever you want based on the AsyncAPI document. You can generate docs and code. It can be used as a library in Nodejs application or through the [AsyncAPI CLI](https://github.com/asyncapi/cli).

The generator tool supports a number of templates to generate code for a variety of different languages and protocols as the output. These templates help to specify what exactly must be generated, and in this tutorial, you use [Nodejs template](https://github.com/asyncapi/nodejs-template).

> 

Reviewer: we also need a kind of prerequisite. 

later in this tutorial user installs MQTT client using npm. Also generated application is Nodejs app that requires installation of dependencies. So we need to make it clear that in this tutorial user as prerequisite needs Nodejs installed

> # Installation Guide

Reviewer: these are correct installation instructions -> https://github.com/asyncapi/generator/pull/818/files#diff-6c85f832c3eae8efc6830b5e9e56020250a891fa7fd58498a8506c4600ab3c34R32-R53

> # Installation Guide

Reviewer: ```suggestion
## Installation Guide
```

Before proceeding to the next step, install the generator tool to use it as a command-line(CLI);

<CodeBlock language="bash">
{`npm install -g @asyncapi/generator`}
</CodeBlock>

The command above helps you install the generator tool globally.

> # Generate code

Reviewer: ```suggestion
## Generate code
```

To generate code from the AsyncAPI document created from the [previous tutorial]("), follow the steps listed below;

1. Trigger generation of the Node.js code:

<CodeBlock language="bash">
{`asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto`}
</CodeBlock>

> The command `asyncapi generate fromTemplate asyncapi.yaml` allowed code to be generated from the AsyncAPI document that was created. The Node.js template was specified by the `@asyncapi/nodejs-template` command.

Reviewer: ```suggestion
The command `asyncapi generate fromTemplate` is how you use AsyncAPI Generator through the AsyncAPI CLI. ` asyncapi.yaml` is how you point to your AsyncAPI document. It can also be a URL. The Node.js template was specified by the `@asyncapi/nodejs-template` command.
```

regarding this and the next parameters you describe.
try instead of 2 paragraphs of text, do a bullet point list like:

```
* `asyncapi generate fromTemplate` is how you use AsyncAPI Generator through the AsyncAPI CLI
* ` asyncapi.yaml` is how you point to your AsyncAPI document. It can also be a URL
* and so on
```

I think it might make it clearer to read. I was thought that if your text can be turned into a bullet point list, you should always do it.

> The `-o` determined where to output the result, and the `-p` defined additional parameters you want to pass to the template, in this instance, the `server`.

Reviewer: ```suggestion
The `-o` determines where to output the result, and the `-p` defines additional parameters you want to pass to the template, in this instance, the `server` parameter is where you specify the server's name as it is defined in AsyncAPI document.
```

2. List all files in directory and check that Node.js application is generated:
<CodeBlock language="bash">
{`cd output && ls`}
</CodeBlock>

> 

Reviewer: what do you think? maybe it would make sense to have here an example of expected result?

like you know: 
You should see something like this in your terminal:
```bash
$ ls
Dockerfile
asyncapi.yaml
docs
src
README.md
config
package.json
```

> # Run your code

Reviewer: ```suggestion
## Start generated application

Now you need to install generated application and start it. At startup time, it will connect with MQTT broker and the topic that was provided in AsyncAPI document.
```

1. Install dependencies of newly generated application:
<CodeBlock language="bash">
{`npm install`}
</CodeBlock>

2. Start the application:
<CodeBlock language="bash">
{`npm start`}
</CodeBlock>

> 

Reviewer: ```suggestion
## Send a message to the broker

Now you will send a message to MQTT broker that generated application is now connected to. You will use command like MQTT client in your terminal.
```

I think we can use the opportunity to improve existing tutorial and add some explanation and clearer structure.
Remember to change steps order

3. In another terminal install the MQTT.js library:
With the MQTT client, which must first be installed, you can create messages.
<CodeBlock language="bash">
{`npm install mqtt -g`}
</CodeBlock>

4. Send a message to your application:
<CodeBlock language="bash">
{`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
</CodeBlock>

> 6. Go back to the previous terminal and check if your application logged the streetlight condition you just sent.

Reviewer: like in case of my comment to `ls` command, we could show here expected result

it is also useful for people that read tutorial but do not execute it on local

## Background context
The [AsyncAPI Generator](https://github.com/asyncapi/generator) is a tool that you can use to generate whatever you want based on the AsyncAPI document. You can generate docs and code. It can be used as a library in a Node.js application or through the [AsyncAPI CLI](https://github.com/asyncapi/cli).

The generator tool supports a number of templates to generate code for a variety of different languages and protocols as the output. These templates help to specify what exactly must be generated, and in this tutorial, you'll use a [Node.js template](https://github.com/asyncapi/nodejs-template).

## Installation guide
<Remember>

It is important to have [Node.js](https://nodejs.org/en/download/) installed which will enable the installation of the required dependencies using <b>npm</b>.

</Remember>

import CliInstallation from '../../../assets/docs/fragments/cli-installation.md' 

<CliInstallation/>

## Generate code

To generate code from the [AsyncAPI document created in a previous tutorial](https://asyncapi.com/docs/tutorials/create-asyncapi-document), follow the steps listed below:

<Remember>

> If you did not follow the previous tutorial and do not have an `asyncapi.yaml` file ready, generate one running `asyncapi new --example=tutorial.yml --no-tty`.
> 
> </Remember>
> 
> 1. Trigger generation of the Node.js code:
>     <CodeBlock language="bash">
>     {`asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto`}
>     </CodeBlock>
>     Let's break down the previous command:
>     - `asyncapi generate fromTemplate` is how you use AsyncAPI Generator via the AsyncAPI CLI. 
>     - ` asyncapi.yaml` is how you point to your AsyncAPI document and can be a URL. 
>     - `@asyncapi/nodejs-template` is how you specify the Node.js template.
>     - `-o` determines where to output the result.
>     - `-p` defines additional parameters you want to pass to the template. Here, the `server` parameter specifies the server's name as it is defined in AsyncAPI document.
> 
> 2. List all files in directory and check that the Node.js application is generated:
> <CodeBlock language="bash">
> {`cd output && ls`}
> </CodeBlock>
> 
> Upon execution of the command above, the following is an example of the expected result:
> ```
> $ ls
> Dockerfile
> asyncapi.yaml
> docs
> src
> README.md
> config
> package.json
> ```
> 
> ## Start generated application
> 1. Install dependencies of the newly generated application:
> <CodeBlock language="bash">
> {`npm install`}
> </CodeBlock>
> 
> 2. Start the application:
> <CodeBlock language="bash">
> {`npm start`}
> </CodeBlock>
> 
> ## Send message to broker
> 1. In another terminal install the MQTT.js library:
>     <CodeBlock language="bash">
> {`npm install mqtt -g`}
>     </CodeBlock>
> 
> 2. Send a message to the MQTT broker that's connected to your generated application. Run the following MQTT client command:
>     <CodeBlock language="bash">
>     {`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
>     </CodeBlock>
> 
> 3. Go back to the previous terminal to check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal:
>     ```bash
>     light/measured was received:
>     { id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }
>     ```
> 

Reviewer: ```suggestion
## Generate code

To generate code from the [AsyncAPI document created in a previous tutorial](https://asyncapi.com/docs/tutorials/create-asyncapi-document), follow the steps listed below:

<Remember>

If you did not follow the previous tutorial and do not have an `asyncapi.yaml` file ready, generate one running `asyncapi new --example=tutorial.yml --no-tty`.

</Remember>

1. Trigger generation of the Node.js code:
    <CodeBlock language="bash">
    {`asyncapi generate fromTemplate asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto`}
    </CodeBlock>

    Let's break down the previous command:
    - `asyncapi generate fromTemplate` is how you use AsyncAPI Generator via the AsyncAPI CLI. 
    - ` asyncapi.yaml` is how you point to your AsyncAPI document and can be a URL. 
    - `@asyncapi/nodejs-template` is how you specify the Node.js template.
    - `-o` determines where to output the result.
    - `-p` defines additional parameters you want to pass to the template. Here, the `server` parameter specifies the server's name as it is defined in AsyncAPI document.

2. List all files in directory and check that the Node.js application is generated:
    <CodeBlock language="bash">
    {`cd output && ls`}
    </CodeBlock>

    Upon execution of the command above, the following is an example of the expected result:
    <CodeBlock language="bash">
    {`$ ls
    Dockerfile
    asyncapi.yaml
    docs
    src
    README.md
    config
    package.json`}
    </CodeBlock>

## Start generated application
1. Install dependencies of the newly generated application:
    <CodeBlock language="bash">
    {`npm install`}
    </CodeBlock>

2. Start the application:
    <CodeBlock language="bash">
    {`npm start`}
    </CodeBlock>

## Send message to broker
1. In another terminal install the MQTT.js library:
    <CodeBlock language="bash">
    {`npm install mqtt -g`}
    </CodeBlock>

2. Send a message to the MQTT broker that's connected to your generated application. Run the following MQTT client command:
    <CodeBlock language="bash">
    {`mqtt pub -t 'light/measured' -h 'test.mosquitto.org' -m '{"id": 1, "lumens": 3, "sentAt": "2017-06-07T12:34:32.000Z"}'`}
    </CodeBlock>

3. Go back to the previous terminal to check if your application logged the streetlight condition you just sent. You should see something like this displayed in the terminal:
    <CodeBlock language="bash">
    {`light/measured was received:
    { id: 1, lumens: 3, sentAt: '2017-06-07T12:34:32.000Z' }`}
    </CodeBlock>

```
